stl/src/filesystem.cpp

```diff
@@ -172,6 +172,16 @@ namespace {
 
         return __std_win_error{GetLastError()};
     }
+
+    [[nodiscard]] bool __stdcall _Is_same_file(_In_ HANDLE _Handle_a, _In_ HANDLE _Handle_b) {
+        constexpr auto _Flags = __std_fs_file_flags::_Backup_semantics | __std_fs_file_flags::_Open_reparse_point;
+        FILE_ID_INFO Id_a, Id_b;
+        if (GetFileInformationByHandleEx(_Handle_a, FileIdInfo, &Id_a, sizeof(FILE_ID_INFO))
+            && GetFileInformationByHandleEx(_Handle_b, FileIdInfo, &Id_b, sizeof(FILE_ID_INFO))) {
+            return memcmp(Id_a.FileId.Identifier, Id_b.FileId.Identifier, sizeof(FILE_ID_128)) == 0;

```

Comment:
A question - how would this behave on simpler filesystems, like `FAT` or `exFAT`. They don't have hard links, but do they have FILE_ID_INFOs? And if not, would `GetFileInformationByHandleEx` fail, or successfully return zero/garbage?

Also looks like `VolumeSerialNumber` should be compared too, there may be different volumes containing files with the same ID

Reply:
`rename` does not work with files across volumes. See  https://pubs.opengroup.org/onlinepubs/009604599/functions/rename.html or https://man7.org/linux/man-pages/man2/rename.2.html. So I don't think we should compare `VolumeSerialNumber`.

With regards to FAT filesystems, I pulled out a USB stick and tried it. Looks like `GetFileInformationByHandleEx` fails with `ERROR_INVALID_PARAMETER` when we request `FileIdInfo`. So no worries 😄